Assignment 4/test/FractionalWrapper.t.sol

```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "yield-utils-v2/mocks/ERC20Mock.sol";
+import "src/FractionalWrapper.sol";
+
+/// @dev In this state, the contract has been initialized and user 1 starts
+///     with some tokens.
+abstract contract ZeroState is Test {
+    event Transfer(address indexed from, address indexed to, uint256 value);
+    event Approval(
+        address indexed owner,
+        address indexed spender,
+        uint256 value
+    );
+
+    event Deposit(
+        address indexed caller,
+        address indexed owner,
+        uint256 assets,
+        uint256 shares
+    );
+    event Withdraw(
+        address indexed caller,
+        address indexed receiver,
+        address indexed owner,
+        uint256 assets,
+        uint256 shares
+    );
+
+    ERC20Mock token;
+    FractionalWrapper wrapper;
+
+    uint256 constant fraction = 7 * (10**26);
+
+    /// @notice The amount of tokens owned by the user.
+    uint256 constant tokensMinted = 1_000_000_000_000_000_000;
+
+    address constant user1 = address(1);
+    address constant user2 = address(2);
+    address constant user3 = address(3);
+
+    function setUp() public virtual {
+        token = new ERC20Mock("Token", "TOK");
+        wrapper = new FractionalWrapper(token, fraction);
+
+        // Give user 1 some tokens
+        token.mint(user1, tokensMinted);
+    }
+}
+
+/// @dev In this state, the user has deposited some tokens.
+abstract contract WrappedTokensMintedState is ZeroState {
+    /// @notice The amount of wrapped tokens minted by depositing assets into
+    ///     the contract.
+    uint256 wrapperTokens;
+
+    function setUp() public override {
+        super.setUp();
+
+        // Deposit some tokens
+        vm.startPrank(user1);
+        token.approve(address(wrapper), tokensMinted);
+        wrapperTokens = wrapper.deposit(tokensMinted, user1);
+        vm.stopPrank();
+    }
+}
+
+contract ZeroStateTest is ZeroState {
+    function testToken() public {
+        assertEq(address(wrapper.token()), address(token));
+    }
+
+    function testAsset() public {
+        assertEq(wrapper.asset(), address(token));
+    }
+
+    function testFraction() public {
+        assertEq(wrapper.fraction(), fraction);
+    }
+
+    function testTotalSupplyAndAssets() public {
+        assertEq(wrapper.totalSupply(), 0);
+        assertEq(wrapper.totalAssets(), 0);
+    }
+
+    function testAllowance(address owner, address spender) public {
+        assertEq(wrapper.allowance(owner, spender), 0);
+    }
+
+    function testBalance(address account) public {
+        assertEq(wrapper.balanceOf(account), 0);
+    }
+
+    function testConvertToShares() public {
```

Comment:
Looks like a fuzzing test would be more convenient.

Reply:
I decided against that as it would basically constitute duplication the function. I thought numbers would be good for testing how the conversions were rounded. I have added a comment explaining this if that's OK.